src/Workspaces/Core/Portable/Workspace/Solution/TextDocumentStates.cs

```diff
@@ -0,0 +1,157 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// Holds on a <see cref="DocumentId"/> to <see cref="TextDocumentState"/> map and an ordering.
+    /// </summary>
+    internal readonly struct TextDocumentStates<TState>
+        where TState : TextDocumentState
+    {
+        public static readonly TextDocumentStates<TState> Empty = new(ImmutableArray<DocumentId>.Empty, ImmutableDictionary<DocumentId, TState>.Empty);
+
+        /// <summary>
+        /// Ordered <see cref="DocumentId"/>s. This is always an <see cref="ImmutableArray{T}"/>,
```

Comment:
❔ How are we avoiding O(n²) complexity for adding many documents?

Reply:
`AddRange` adds multiple documents.